pykafka/rdkafka/producer.py

```diff
@@ -0,0 +1,133 @@
+import logging
+
+from pykafka.producer import Producer, CompressionType
+from pykafka.handlers import ResponseFuture
+from . import _rd_kafka
+
+
+log = logging.getLogger(__name__)
+
+
+class RdKafkaProducer(Producer):
+    """A librdkafka-backed version of pykafka.Producer
+
+    This aims to conform to the Producer interface as closely as possible.  One
+    notable difference is that it returns a future from produce(), through
+    which the user can later check if produced messages have made it to kafka.
+
+    For an overview of how configuration keys are mapped to librdkafka's, see
+    _mk_rdkafka_config_lists.
+    """
+    def start(self):
+        if not self._running:
+            brokers = ','.join(map(lambda b: ':'.join((b.host, str(b.port))),
+                                   self._cluster.brokers.values()))
+            conf, topic_conf = self._mk_rdkafka_config_lists()
+
+            self._rdk_producer = _rd_kafka.Producer()
+            self._rdk_producer.configure(conf=conf)
+            self._rdk_producer.configure(topic_conf=topic_conf)
+            self._rdk_producer.start(brokers, self._topic.name)
+            self._running = True
+
+            # TODO spawn thread to regularly poll() for delivery reports (and
+            # break if not _running)
+
+    def _produce(self, message_partition_tup):
+        # Delivery-reports will be available through this future
+        future = ResponseFuture(self._cluster.handler)
+
+        (key, msg), part_id = message_partition_tup
+        self._rdk_producer.produce(msg, key, part_id, future)
+        return future
```

Comment:
I guess the reason to do this is to take advantage of rdkafka's delivery reports, which makes sense. Implementing these in the `Producer` should be straightforward, since we're checking for message send success anyway. I believe I'd do it by having `OwnedBroker.enqueue` create and return a future that it also adds to the enqueued `(kv, partition_id)` tuple. On the other side, we could fill in the future with whatever values were appropriate based on the fate of the message.

Given that I can picture the implementation already, there's a clear upside (more descriptive return values from `produce`) and I can't come up with any huge downsides for the average user (who can simply ignore that `produce` returns anything), I'm +1 to this idea. The question in my mind is whether there is any better way to provide message statuses to the client - a callback interface sounds clumsy, and we already know that a synchronously returned status from `produce` only gets us half of the way there.

Assuming we agree this makes sense, it should have its own issue number so we can more easily partition the pure python and rdkafka implementations.